index.js

```diff
@@ -1,189 +1,317 @@
+ * @api private
  */
 
-function getPathValue(parsed, obj) {
-  var tmp = obj;
-  var res;
+function internalGetPathValue(parsed, obj, pathDepth) {
+  var temporaryValue = obj;
+  var res = null;
+  pathDepth = (typeof pathDepth === 'undefined' ? parsed.length : pathDepth);
 
-  for (var i = 0, l = parsed.length; i < l; i++) {
+  for (var i = 0; i < pathDepth; i++) {
     var part = parsed[i];
-    if (tmp) {
-      if (defined(part.p)) tmp = tmp[part.p];
-      else if (defined(part.i)) tmp = tmp[part.i];
-      if (i == (l - 1)) res = tmp;
-    } else {
-      res = undefined;
+    if (temporaryValue) {
+      if (typeof part.p === 'undefined') {
+        temporaryValue = temporaryValue[part.i];
+      } else {
+        temporaryValue = temporaryValue[part.p];
+      }
+
+      if (i === (pathDepth - 1)) {
+        res = temporaryValue;
+      }
     }
   }
 
   return res;
-};
+}
+
+/* !
+ * ## defined(val)
+
+ * Check if `val` is defined.
+ *
+ * @param {Mixed} val
+ * @returns {Boolean} `true` if defined
+ * @api private
+ */
+
+function defined(val) {
+  return !(!val && typeof val === 'undefined');
+}
 
-/*!
+/* !
+ * ## internalSetPathValue(parsed, value, obj)
+ *
  * Companion function for `parsePath` that sets
  * the value located at a parsed address.
  *
- *  setPathValue(parsed, 'value', obj);
+ *  internalSetPathValue(parsed, 'value', obj);
  *
  * @param {Object} parsed definition from `parsePath`
  * @param {*} value to use upon set
  * @param {Object} object to search and define on
  * @api private
  */
 
-function setPathValue(parsed, val, obj) {
-  var tmp = obj;
-  var i = 0;
-  var l = parsed.length;
-  var part;
-
-  for (; i < l; i++) {
+function internalSetPathValue(parsed, val, obj) {
+  var tempObj = obj;
+  var pathDepth = parsed.length;
+  var part = null;
+  // Here we iterate through every part of the path
+  for (var i = 0; i < pathDepth; i++) {
+    var propName = null;
+    var propVal = null;
     part = parsed[i];
 
-    if (defined(tmp) && i == (l - 1)) {
-      var x = defined(part.p) ? part.p : part.i;
-      tmp[x] = val;
-    } else if (defined(tmp)) {
-      if (defined(part.p) && tmp[part.p]) {
-        tmp = tmp[part.p];
-      } else if (defined(part.i) && tmp[part.i]) {
-        tmp = tmp[part.i];
-      } else {
-        var next = parsed[i + 1];
-        var x = defined(part.p) ? part.p : part.i;
-        var y = defined(next.p) ? {} : [];
-        tmp[x] = y;
-        tmp = tmp[x];
-      }
+    // If it's the last part of the path, we set the 'propName' value with the property name
+    if (defined(tempObj) && i === (pathDepth - 1)) {
+      propName = defined(part.p) ? part.p : part.i;
+      // Now we set the property with the name held by 'propName' on object with the desired val
+      tempObj[propName] = val;
+    } else if (defined(tempObj) && defined(part.p) && tempObj[part.p]) {
+      tempObj = tempObj[part.p];
+    } else if (defined(tempObj) && defined(part.i) && tempObj[part.i]) {
```

Comment:
Yeah, you are right, since we already got this condition on the clause `tempObj[part.p]` we won't need `defined(tempObj)`. The only way it would be undefined was if the object passed to the function was `undefined`, and in that case it would fail anyway.
I'll remove it today and `amend` into this commit.
Thanks for noticing.